File: ROOT_Macro/csvToTTreeConverter/csvToTree.py
#!/usr/bin/python3
import os
from sys import exit
import csv
import numpy as np
from datetime import datetime
from argparse import ArgumentParser
from ROOT import TFile, TTree, TMath

start = datetime.now()

words_in_line = 54
glat_word_counter = 23
thetas_word_counter = 19

def main(args=None):
	parser = ArgumentParser(usage="Usage: %(prog)s [options]", description="CSV to TTree converter")
	parser.add_argument("-l", "--list", type=str,dest='list', help='Input CSV DATA list')
	parser.add_argument("-o", "--output", type=str,dest='output', help='output TFile')
	parser.add_argument("-v", "--verbose", dest='verbose', default=False,action='store_true', help='run in high verbosity mode')
	opts = parser.parse_args(args)
	
	# Create glon e glat variables
	glat = np.empty((1), dtype="float")
	glon = np.empty((1), dtype="float")
	thetas = np.empty((1), dtype="float")
	phi = np.empty((1), dtype="float")
	events = np.empty((1), dtype="int")
	geo_lat = np.empty((1), dtype="float")
	geo_lon = np.empty((1), dtype="float")

	# Create new TFile
	RTI_output_file = TFile(opts.output, "RECREATE")
	if RTI_output_file.IsZombie():
		print('Error writing output TFile [{}]'.format(opts.output))
		exit()

	# Create final Tree
	RTI_tree = TTree("RTI_tree", "AMS RTI TTree")

	# Branch the TTree
	RTI_tree.Branch("glat", glat, "glat/D")
	RTI_tree.Branch("glon", glon, "glon/D")
	RTI_tree.Branch("thetas", thetas, "thetas/D")
	RTI_tree.Branch("phi", phi, "phi/D")
	RTI_tree.Branch("events", events, "events/I")
	RTI_tree.Branch("geo_lat", geo_lat, "geo_lat/D")
	RTI_tree.Branch("geo_lon", geo_lon, "geo_lon/D")
	
	# Parse input list
	with open(opts.list, 'r') as input_list:
		csv_files = input_list.readlines()
		for file in csv_files:
			with open(file.rstrip("\n"), 'r') as tmp_file:
				csv_reader = csv.reader(tmp_file, delimiter=' ', quoting=csv.QUOTE_NONE)
				line_count = 0
				for row in csv_reader:
					if line_count != 0:
						if int(row[1]):
							thetas[0] = float(row[18])
							phi[0] = float(row[19])
							glat[0] = float(row[22])
							glon[0] = float(row[23])
							events[0] = int(row[24])
							geo_lat[0] = thetas*TMath.RadToDeg()
							geo_lon[0] = phi*TMath.RadToDeg()
							RTI_tree.Fill()
					line_count += 1
				if opts.verbose:
					print('Has been read {} lines [{}]'.format(line_count, file.rstrip("\n")))
	RTI_tree.Write()
	RTI_output_file.Close()

if __name__ == "__main__":
	main()
	print(datetime.now()-start)

File: ROOT_Macro/csvToTTreeConverter/csvToTree.cpp
#include <iostream>
#include <stdlib.h>
#include <stdio.h>
#include <string>
#include <fstream>
#include <sstream>

#include "TTree.h"
#include "TFile.h"
#include "TMath.h"

#define words_in_line 54
#define glat_word_counter 23
#define thetas_word_counter 19

std::string readStreamFromFile(std::string input_RTI_path)
{
    std::ifstream input_file(input_RTI_path.c_str());
    if (!input_file.is_open())
    {
        std::cerr << "\nERROR 100! File not open " << input_RTI_path << "\n\n";
        exit(100);
    }
    std::string input_string((std::istreambuf_iterator<char>(input_file)), (std::istreambuf_iterator<char>()));
    input_file.close();
    return input_string;
}

void parse_stream(
    const std::string input_RTI_path,
    double &glat, 
    double &glon,
    double &thetas,
    double &phi,
    int &events,
    double &geo_lat,
    double &geo_lon,
    TTree &RTI_tree)
{
    std::istringstream input_stream(readStreamFromFile(input_RTI_path));
    std::string tmp_RTI_file;

    while (input_stream >> tmp_RTI_file)
    {
        std::istringstream rti_stream(readStreamFromFile(tmp_RTI_file));
        // Create line counter variable
        unsigned int line_counter = 0;
        
        std::string tmp_str;
        std::string header;
        std::string::size_type sz;

        // parse stream
        //while (input_stream >> tmp_str)
        rti_stream >> header;
        int wcounter = 0;

        while(rti_stream >> tmp_str)
        {
            ++wcounter;

            if (wcounter == 1)
                auto JMDCtime = stoul(tmp_str, &sz);
            else if (wcounter == 2)
            {
                auto run = stoul(tmp_str, &sz);
                if (!run)
                {
                    ++line_counter;
                    wcounter=0;
                    continue;
                }
            }
            else if (wcounter == thetas_word_counter)
                thetas = stod(tmp_str, &sz);
            else if (wcounter == (thetas_word_counter+1))
                phi = stod(tmp_str, &sz);
            else if (wcounter == glat_word_counter) 
                glat = stod(tmp_str, &sz);
            else if (wcounter == (glat_word_counter+1))
                glon = stod(tmp_str, &sz);
            else if (wcounter == (glat_word_counter+2))
                events = stoi(tmp_str, &sz);
            else if (wcounter == words_in_line)
            {
                wcounter=0;
                ++line_counter;
                geo_lat = thetas*TMath::RadToDeg();
                geo_lon = phi*TMath::RadToDeg();
                RTI_tree.Fill();
            }
        }
        
        std::cout << "\nHas been read " << line_counter << " lines from: " << tmp_RTI_file;
    }
}

void csvToTree(std::string input_RTI_path, std::string outFile)
{
    // Create glon e glat variables
    double glat = -1;
    double glon = -1;
    double thetas = -1;
    double phi = -1;
    int events = -1;
    double geo_lat = -1;
    double geo_lon = -1;

    // Create new TFile
    TFile RTI_output_file(outFile.c_str(),"RECREATE");
    if (RTI_output_file.IsZombie())
    {
        std::cerr << "\n\nError writing output TFile: " << outFile << std::endl;
        exit(100);
    }

    // Create final Tree
    TTree RTI_tree("RTI_tree", "AMS RTI TTree");
    // Branch the TTree
    RTI_tree.Branch("glat", &glat, "glat/D");
    RTI_tree.Branch("glon", &glon, "glon/D");
    RTI_tree.Branch("thetas", &thetas, "thetas/D");
    RTI_tree.Branch("phi", &phi, "phi/D");
    RTI_tree.Branch("events", &events, "events/I");
    RTI_tree.Branch("geo_lat", &geo_lat, "geo_lat/D");
    RTI_tree.Branch("geo_lon", &geo_lon, "geo_lon/D");

    // Parse stream
    parse_stream(
        input_RTI_path,
        glat, 
        glon,
        thetas,
        phi,
        events,
        geo_lat,
        geo_lon,
        RTI_tree);
    
    RTI_tree.Write();
}